WcaOnRails/app/models/competition.rb

```diff
@@ -622,6 +622,62 @@ def continent
     country ? Continent.find_by_id(country.continentId) : nil
   end
 
+  def psych_sheet_event(event)
+    preferred_format = event.preferred_formats.first
+    sort_clause = (preferred_format.sort_by == :single) ? "-single_rank desc, -average_rank desc, users.name" : "-average_rank desc, -single_rank desc, users.name"
+
+    joinsql = <<-ENDSQL
+      join users on users.id = Preregs.user_id
+      join Countries on Countries.iso2 = users.country_iso2
+      left join RanksSingle on RanksSingle.personId = users.wca_id and RanksSingle.eventId = '#{event.id}'
+      left join RanksAverage on RanksAverage.personId = users.wca_id and RanksAverage.eventId = '#{event.id}'
+    ENDSQL
+
+    selectsql = <<-ENDSQL
+      Preregs.id,
+      users.name select_name,
+      users.wca_id select_wca_id,
+      Preregs.accepted_at,
+      Countries.name select_country,
+      registration_events.event_id,
+      RanksAverage.worldRank average_rank,
+      ifnull(RanksAverage.best, 0) average_best,
+      RanksSingle.worldRank single_rank,
+      ifnull(RanksSingle.best, 0) single_best
+    ENDSQL
+
+    registrations = self.registrations.
+                         accepted.
+                         joins(joinsql).
+                         joins(:registration_events).
+                         where("registration_events.event_id=?", event.id).
+                         order(sort_clause).
+                         select(selectsql)
+
+    prev_registration = nil
+    registrations.each_with_index do |registration, i|
+      registration.tied_previous = false
+      if prev_registration
+        registration.tied_previous = if preferred_format.sort_by == :single
+                                       registration.single_rank == prev_registration.single_rank
+                                     else
+                                       registration.average_rank == prev_registration.average_rank
+                                     end
+        registration.pos = registration.tied_previous ? prev_registration.pos : i + 1
+      else
+        registration.pos = i + 1
+      end
+      has_competed = if preferred_format.sort_by == :single
+                       registration.single_rank
+                     else
+                       registration.average_rank
+                     end
+      registration.pos = nil unless has_competed
+      prev_registration = registration
+    end
```

Comment:
@pingskills okay I played with this a bit more because I believed this could be really improved.
Firstly: my version works as fast as the original *for me*, I've no idea what's the difference.
Also I've found a way to optimize this even more by stopping computation when we stumble upon a registration that has never competed (as we sort by rank, all subsequent ones hasn't too).

[Here are three methods](http://pastebin.com/4x9R8PCq) I added to Competition model:
- the original (`psych_sheet_event`)
- modified without send (`psych_sheet_event_v2`)
- modified with send (`psych_sheet_event_v3`)

(The part that differs is after assigning `registrations`.)

I've tested them for competition with over 600 accepted registrations for 3x3x3, 20 of them has never competed in it. The code I used for the benchmark is [here](http://pastebin.com/NcKRirV8).
Result (tested in console):
```
original: 0.09891407200999311 (avg of 100)
modified without send: 0.08930202386000019 (avg of 100)
modified with send: 0.0989089906800109 (avg of 100)
```

Of course I've checked that the returned `registrations` are the same for each of these three methods.

I'd be happy with either v2 of v3. I hope that's pretty clean now =)